--- a/board/renesas/rzv2l-dev/rzv2l-dev.c
+++ b/board/renesas/rzv2l-dev/rzv2l-dev.c
@@ -44,9 +44,9 @@
 #define CPG_CLKON_USB				(CPG_BASE + 0x578)
 
 /* PFC */
-#define PFC_P37						(PFC_BASE + 0x037)
-#define PFC_PM37					(PFC_BASE + 0x16E)
-#define PFC_PMC37					(PFC_BASE + 0x237)
+#define PFC_P38					(PFC_BASE + 0x038)
+#define PFC_PM38					(PFC_BASE + 0x170)
+#define PFC_PMC38					(PFC_BASE + 0x238)
 #define PFC_PWPR 					(PFC_BASE + 0x3014)
 #define PFC_PMC14					(PFC_BASE + 0x214)
 #define PFC_PFC14					(PFC_BASE + 0x450)
@@ -72,15 +72,10 @@
 void s_init(void)
 {
 
-	/* SD1 power control: P39_1 = 0; P39_2 = 1; */
-	*(volatile u32 *)(PFC_PMC37) &= 0xFFFFFFF9; /* Port func mode 0b00 */
-	*(volatile u32 *)(PFC_PM37) = (*(volatile u32 *)(PFC_PM37) & 0xFFFFFFC3) | 0x28; /* Port output mode 0b1010 */
-#if CONFIG_TARGET_SMARC_RZV2L
-	*(volatile u32 *)(PFC_P37) = (*(volatile u32 *)(PFC_P37) & 0xFFFFFFF9) | 0x6;	/* Port 39[2:1] output value 0b11*/
-#else
-
-	*(volatile u32 *)(PFC_P37) = (*(volatile u32 *)(PFC_P37) & 0xFFFFFFF9) | 0x4;	/* Port 39[2:1] output value 0b10*/
-#endif
+	/* SD1 power control for MXT SMARC SOM power enable: P40_2 = 1; */
+	*(volatile u32 *)(PFC_PMC38) &= 0xFFFFFFF8; /* Port GPIO mode 0b000 */
+	*(volatile u32 *)(PFC_PM38) = (*(volatile u32 *)(PFC_PM38) & 0xFFFFFFC0) | 0x20; /* Port output mode 0b100000 */
+	*(volatile u32 *)(PFC_P38) = (*(volatile u32 *)(PFC_P38) & 0xFFFFFFF8) | 0x4;	/* Port 40[2] output value 0b100*/
 
 	/* can go in board_eht_init() once enabled */
 	*(volatile u32 *)(ETH_CH0) = (*(volatile u32 *)(ETH_CH0) & 0xFFFFFFFC) | ETH_PVDD_1800;
